### ISSUE TITLE
Compilation mode regexps

### ISSUE THREAD

**Issue by dbuenzli** [ocaml/caml-mode#5]
It seems that caml-mode installs it's own compilation mode regexps [here](https://github.com/ocaml/caml-mode/blob/38ebde12d3d529e6ef8078967997d32226e69e82/caml.el#L790-L831). 

After `caml-mode` is loaded this seems to break the recognition made nowadays by default by compilation mode, see [this regexp]( https://github.com/emacs-mirror/emacs/blob/b7a98993789d18bc675798d49038982d5cf41683/lisp/progmodes/compile.el#L163-L166).

More specifically multi-line comments in the style of that regexp no longer get recognized. I bet this may be due to an ordering issue.

I'm a little bit unsure what should be done, maybe no longer install these regexps ?

**Comment by Chris00** [comment]
The compilation regexp is [being worked on for Tuareg](https://github.com/Chris00/tuareg/blob/master/tuareg.el#L2647).  Note that this regexp is compatible with OCaml 4.08.  Ultimately, I'm thinking of using the same regexp for both modes — installing it only once of course.

**Comment by dbuenzli** [comment]
I think the question is rather why install one at all ? 

The problem here is that installing a new regexp is a side effect. In my case I'm using a format that is recognized by *compilation* mode by default. But once I have a buffer that loads `caml-mode` that format is no longer recognized.

**Comment by Chris00** [comment]
On 15 April 2019 at 00:09 CEST, Daniel Bünzli wrote:
>
> I think the question is rather why install one at all ?
>
> The problem here is that installing a new regexp is a side effect. In my case I'm using a format that is recognized by compilation mode by default. But once I have a buffer that loads caml-mode that format is no longer recognized.

If 4.08 did not change the way errors were reported, I think there would be no need to install a new one.  But it does and it is desirable to distinguish errors from warnings.  If we keep the old regexp of caml-mode, they should be pushed to the end of the list in order not to interfere (I do not think they are relevant anymore, I would be in favor of dropping them).

**Comment by dbuenzli** [comment]
> The problem here is that installing a new regexp is a side effect. In my case I'm using a format that is recognized by _compilation_ mode by default. But once I have a buffer that loads `caml-mode` that format is no longer recognized.

It's a bit odd. I no longer see that difference (but I did update my emacs meanwhile). However it seems both regexpes have a problem in that they interpret `characters X1-X2` as one-based *column* numbers while they should be interpreting X1-X2 as *positions* or insertion points (slits between columns). For example with the following error report:
```shell
> ocamlc -c a.ml
File "a.ml", line 1, characters 11-14:
1 | function x let x
               ^^^
Error: Syntax error
```
Both `compilation-mode` and `compilation-mode` extended by `caml-mode` highlight the space before the let and the `let`. This means the regexp interprets 11-14 as one-based columns rather than positions. Graphically (above columns, below position):

```
                     -------
                   0 1 2 3 4 5 6                  
 1 2 3 4 5 6 7 8 9 1 1 1 1 1 1
|f|u|n|c|t|i|o|n| |x| |l|e|t| |x|
0 1 2 3 4 5 6 7 8 9 1 1 1 1 1 1 1
                    0 1 2 3 4 5 6
                      -------
```

@Chris00 any idea how to move on to fix this ?

**Comment by dbuenzli** [comment]
Ping @Chris00.

**Comment by dbuenzli** [comment]
Another issue is that both the compilation mode default and caml-mode do not recognize [ocaml backtraces](https://github.com/ocaml/ocaml/commit/b7716c46256776f4af55ac460bf7bdba26c0005c). I used to have that working at some point but it's always a bit hard to know pin down what changed, between my `.emacs` cleanup, emacs changes, `caml-mdoe` and compiler changes…

In any case tweaking [compilation's mode default](https://github.com/emacs-mirror/emacs/blob/b7a98993789d18bc675798d49038982d5cf41683/lisp/progmodes/compile.el#L163-L166) to:
```
(add-to-list 'compilation-error-regexp-alist-alist
 '(ocaml
   "^.*[Ff]ile \\(\"?\\)\\([^,\" \n\t<>]+\\)\\1, lines? \\([0-9]+\\)-?\\([0-9]+\\)?\\(?:$\\|,\
\\(?: characters? \\([0-9]+\\)-?\\([0-9]+\\)?:\\)?\\([ \n]Warning\\(?: [0-9]+\\)?:\\)?\\)"
   2 (3 . 4) (5 . 6) (7)))
(add-to-list 'compilation-error-regexp-alist 'ocaml)
```
Seems to do it for me. 

Shouldn't we try to maybe have that fixed in emacs upstream and cleanup both `caml-mode` and `tuareg` ? I'm terrible both with regexps and emacs programming but I'm willing to give it a hand if there's a plan.

**Comment by Chris00** [comment]
I've used the regexp in Tuareg (which was ported to `rx` by @mattiase).  [Tested in `compilation.txt`](https://github.com/ocaml/tuareg/blob/master/compilation.txt).  @dbuenzli Shortly after the next release of caml-mode/Tuareg, I'll submit a patch to Emacs.

**Comment by dbuenzli** [comment]
Excellent @Chris00 thanks so much for this ! Should we have maybe someone from upstream have a second look at the regexp before it gets into Emacs ?

**Comment by Chris00** [comment]
@dbuenzli Sure, more eyeballs are always better.

**Comment by dbuenzli** [comment]
I don't know who uses emacs upstream maybe @gasche can suggest someone ? 

Also just for reference here is the PR that changed the backtraces https://github.com/ocaml/ocaml/pull/9096 but I see you actually took care to test both pre and post 4.11.

**Comment by gasche** [comment]
I'm an Emacs user, but I haven't tested the new regexp. I would trust the testing more than myself. Reading the regexp, I'm finding out about [rx structured regexps](https://www.gnu.org/software/emacs/manual/html_node/elisp/Rx-Notation.html) which is indeed nice. I don't know if mixing backtrace regexps and error regexps is really ideal for readability, but I guess that's okay.

One small aspect I wondered is why there is some logic to detect "Warning" at the end (after the location part), but not "Error" (and maybe "Alert" which one can also get these days, I think?). My vague understanding from looking at the [obscure documentation](http://doc.endlessparentheses.com/Var/compilation-error-regexp-alist.html) is that the TYPE part is set just so that returning `nil` means error and `1` means Warning, and the current behavior of a group within an optional part of the regexp behaves like that by chance. I think this could be worth a comment in the code.

There is logic in the regexp to parse the `3 |   let foo = bar` code-quotation format, but I don't know if it is robust to other source-highlighting styles (in particular the legacy style with `-error-style short`), and I'm not sure how easy it is to get those style used from within Emacs -- probably not too hard if the user asks for them in their build script for some reason. This part of the regexp again would deserve a comment, and I would think about its robustness wrt. other location-reporting styles.

**Comment by mattiase** [comment]
Note that the auxiliary function `tuareg--end-column` is required for this to work, since it's referenced from the entry added to `compilation-error-regexp-alist-alist`. You probably want to rename that to something else in caml-mode.

I'd be happy to help you get the message regexps you need into Emacs, but there are a couple of items worth considering:

First, the release pace of Emacs is, well, stately. Years go between major releases; OCaml is very nimble in comparison. Users will need external packages to supplement the built-in message matching rules in any case. This is especially true given that many Emacs users are slow to upgrade. Thus, updating the rules in Emacs core does not remove the need for them in tuareg, caml-mode etc, which can be adjusted in a blink.

There is also a small matter of compatibility – Emacs users may customise the `compilation-error-regexp-alist` variables to contain a selection of rules, so we can't go on renaming them or substantially alter their scope. More specifically, the one named `python-tracebacks-and-caml` probably needs to stay put but we could make it more useful, and add another one.

**Comment by Chris00** [comment]
> Note that the auxiliary function tuareg--end-column is required 

Of course (fixed by a95aaca)!  Too tired, will stop for today!

**Comment by mattiase** [comment]
@gasche, regarding using the same regexp for both backtraces and messages, that was done for pragmatic reasons (ie, laziness): a lot of the messy parts are common to both. They could be pried apart if you prefer that – it would allow users to use them independently, but it's unclear how that would be an actual benefit.

I think a range of more-or-less legacy message styles are covered by the regexp. If you can come up with some notable examples that aren't represented in `tuareg-tests.el`, tell me and I'll prepare a PR.

Emacs only has the message levels Error, Warning and Info. If OCaml says Alert, which would that be? Right now they are probably counted as Error by default, but I'm guessing that Warning would be more appropriate since it's mainly about deprecation, old syntax and so on?

**Comment by dbuenzli** [comment]
Thanks @gasche for having a peek at it. 

@mattiase you are of course right. The situation is not ideal. 

The reason why I'd like to have it in upstream emacs is that it's not uncommon to invoke compilation mode before you actually load  `{caml,tuareg}-mode`, for example if you open your build system configuration file and invoke the build. 

In particular it would be nice if we could avoid the kind of [situation](https://github.com/ocaml/caml-mode/issues/5#issue-432960752) that prompted me to open this issue.

**Comment by mattiase** [comment]
I agree that it would make sense to provide Emacs with some more up-to-date rules for OCaml, even if Tuareg & co then override those with their potentially more up-to-date rules.

The question is what we then do about the `python-tracebacks-and-caml` rule in Emacs. I'm not overly fond of these multi-purpose rules: they are subject to the whims of several independent parties and with time tend to become inadequate for all. It would be probably be a good idea to tear it apart, but it's a problem when their matched sets intersect.

In theory they don't: Python tracebacks start with exactly two spaces before the `File` part, OCaml compiler messages with zero spaces (seven spaces for ancillary locations; we should probably treat those as Info-level messages).

@monnier, you wrote that rule in response to Emacs bug https://debbugs.gnu.org/cgi/bugreport.cgi?bug=8585 which aimed to satisfy Buildbot that prefixes Python messages with a tab. What about we make a `python-traceback` rule which starts with an optional tab followed by two spaces, and use the Tuareg rule for OCaml, tweaked so that it doesn't intersect? We could keep the `python-tracebacks-and-caml` rule on the books for compatibility but just not include it in `compilation-error-regexp-alist` by default.

It's quite likely that Buildbot's tab-prefixing isn't in any way specific to Python messages but applied to anything that passes through its digestive tract, so adding optional spaces to a specific rule isn't really congruent. It may be better to generalise the OMake hackery and let users customise a prefix pattern.

**Comment by monnier** [comment]
I don't have a strong opinion on this.  Your analysis sounds fine to me.

The only opinion I have in this area is that the better answer is to fix
OCaml to give messages in the standard GNU convention format so that it
"just works".

**Comment by mattiase** [comment]
>  the better answer is to fix
> OCaml to give messages in the standard GNU convention format so that it
> "just works".

No dispute from me, but I assumed that there were reasons good and bad for the current message format...

My previous claim that OCaml's ancillary locations are indented by exactly 7 spaces is neither quite nor necessarily true: it's just emergent behaviour from a somewhat over-engineered implementation. For errors, it's often sort of true because there's a hanging indent determined by the width of `"Error: "`, but for warnings and alerts the indentation is more likely to be zero.

This isn't really satisfactory in the machine-readability department. Marking the ancillary locations as Info (instead of Error) feels right since those shouldn't contribute to Emacs's error count, and it's nice to get them painted in a different colour. However, Info locations are ignored by `next-error`/`previous-error` unless the user customises `compilation-skip-threshold` which most people don't know about.

What does the community think – should those extra locations be of Error or Info? (Warnings rarely have them so I'm conveniently ignoring them here.)

@dbuenzli made a good point – wouldn't it be nice if the compilation rules could be extracted into autoloads, so that the tuareg or caml-mode packages don't need to be loaded for the compilation errors to be parsed correctly? Maybe doing so breaks all rules (which naturally fills me with delight). Many other programming modes must have the same problem.

**Comment by monnier** [comment]
> No dispute from me, but I assumed that there were reasons good and bad for
> the current message format...

No doubt.  IIRC it's been requested and rejected in the past already.
I just mentioned this to give some context to my position.

> What does the community think – should those extra locations be of Error or
> Info? (Warnings rarely have them so I'm conveniently ignoring them here.)

I vote for "info" because I think it's the only choice which makes
sense: they are not errors.

> @dbuenzli made a good point – wouldn't it be nice if the compilation rules
> could be extracted into autoloads, so that the tuareg or caml-mode packages
> don't need to be loaded for the compilation errors to be parsed correctly?

We could do that, indeed.  I don't forsee any particular issue
except for the need to wrap the `add-to-list` code into a
`with-eval-after-load`.

**Comment by mattiase** [comment]
> I vote for "info" because I think it's the only choice which makes sense: they are not errors.

Good, that is now PR ocaml/tuareg#266.

> We could do that, indeed.  I don't forsee any particular issue except for the need to wrap the `add-to-list` code into a `with-eval-after-load`.

It seems to work. I have a patch for tuareg.

**Comment by dbuenzli** [comment]
> No doubt. IIRC it's been requested and rejected in the past already. I just mentioned this to give some context to my position.

@monnier this is a bit tangential to the question but do you know if there's any work to update these standards to be made to work with Unicode text ? 

With Unicode, asking cli tools to compute columns without having knowledge of the rendering layer is totally doomed (`wcwidth` included). 

What cli tools can easily give is either the number of unicode scalar values or the number of bytes in the encoding since the beginning of line.

**Comment by mattiase** [comment]
The [GNU coding standards](https://www.gnu.org/prep/standards/html_node/Errors.html) say:

> For non-ASCII characters, Unicode character widths should be used when in a UTF-8 locale; GNU libc and GNU gnulib provide suitable wcwidth functions.

which as you point out isn't always practical:

> With Unicode, asking cli tools to compute columns without having knowledge of the rendering layer is totally doomed (wcwidth included).

Agreed, counting columns (which is what Emacs does by default) is sort of a dead end because of this. No two terminals completely agree with one another or with non-TTY Emacs.

Compilers that output source lines with a caret to indicate the position have the same problem.

**Comment by dbuenzli** [comment]
Yes output to terminal will remain problematic whatever happens – some details written by @pqwy are available in the long doc string of [that function](https://erratique.ch/software/uucp/doc/Uucp/Break/index.html#terminal_width) which provides a kind of `wcwidth` for OCaml.

But for editors it doesn't have to be since they actually have both the text data and knowledge of the rendering layer. Keeping it the simplest for cli tools, that is not even ask them to consult Unicode character properties or seemingly working but broken derivatives like `wcwidth` seems a better way forward. 

(At some point I thought I had found a way by using `(setq compilation-error-screen-columns nil)` but either I dreamed it worked or it stopped working at some point)

**Comment by mattiase** [comment]
> (At some point I thought I had found a way by using (setq compilation-error-screen-columns nil) but either I dreamed it worked or it stopped working at some point)

Yes, that variable seems to be broken – will investigate.

**Comment by monnier** [comment]
> @monnier this is a bit tangential to the question but do you know if there's
> any work to update these standards to be made to work with Unicode text ?
> With Unicode, asking cli tools to compute columns without having knowledge
> of the rendering layer is totally doomed (`wcwidth` included).

I don't think this is being worked on, sadly.  I think the current spec
is indeed making a bad choice (my own favorite is to count what you call
Unicode scalar values, tho bytes would be OK as well).

This said, the current bad choice is not completely doomed, if you
consider `wcwidth` as the definition of the width (both in the reporting
tool and in the tool interpreting the report): it doesn't give the same
column position as a rendering layer but it does correctly convey the
info about the actual position in the text (plus or minus a few
zero-width unicode values, admittedly).

It's just more cumbersome and less precise than counting unicode
values :-(

In any case, it's offtopic here.  I'd encourage you to bring it up to
the GNU folks at ***@***.*** (and feel free to put me in Cc:
if you want my support).

**Comment by mattiase** [comment]
> Yes, [compilation-error-screen-columns] seems to be broken – will investigate

No, it seems to work. The problem with OCaml is of course that its columns are counted in neither columns nor Unicode scalar values but bytes, for which Emacs has no setting. We could add one, assuming that changing the compiler is out of the question.

**Comment by dbuenzli** [comment]
Wasn't that supposed to be fixed ? 

In emacs 28.2, the wrong [highlight here](https://github.com/ocaml/caml-mode/issues/5#issuecomment-535001997) still happens with `compilation-mode` whether `caml-mode` (v4.9) is used or not.

(Since I'm using the format in another context I'd happily switch to gnu's convention but in contrast to ocaml's format it's not possible to indent locations, it thinks you have file paths that starts with spaces if you do :-()

**Comment by dbuenzli** [comment]
So I'm not sure I understand the [logic here](https://github.com/ocaml/caml-mode/blob/a970f303065fe176a920db0466dacf3e2a20b56b/caml.el#L755-L767) about emacs 28. 

Also the match still doesn't seem to take into account that the start character is zero-based in OCaml messages while the regexp should return one-based columns if I'm not mistaken. 

I tried to adjust the start column with a `caml--begin-column` in my `.emacs`, it seems to do the trick. 

```
(require 'compile)
(require 'caml)

(defun caml--begin-column ()
  (and (match-beginning 6)
       (+ (string-to-number (match-string 6)) 1)))

(defun caml--end-column ()
  (and (match-beginning 7)
       (string-to-number (match-string 7))))

(when (boundp 'compilation-error-regexp-alist-alist)
  (push `(ocaml ,caml--error-regexp 3 (4 . 5)
                (caml--begin-column . caml--end-column) (8) 1
                (8 font-lock-function-name-face))
        compilation-error-regexp-alist-alist))

(when (boundp 'compilation-error-regexp-alist)
  (push 'ocaml compilation-error-regexp-alist))
```

I doubt however this or what is in master correctly handles multi-line ranges.

**Comment by dbuenzli** [comment]
> I doubt however this or what is in master correctly handles multi-line ranges.

Actually maybe it's right since https://github.com/ocaml/ocaml/pull/8541 I was no longer sure what the end character was and upstream doesn't seem interested in documenting the format (https://github.com/ocaml/ocaml/issues/8978).

**Comment by mattiase** [comment]
Looks like caml-mode doesn't set `compilation-first-column` (tuareg does, and `compilation-error-screen-columns`). It probably should.

(Any reason for preferring caml-mode to tuareg?)

**Comment by dbuenzli** [comment]
> Looks like caml-mode doesn't set `compilation-first-column` (tuareg does, and `compilation-error-screen-columns`). It probably should.

Wasn't the idea to upstream the regexps ? In that case it would be better if they worked without having to fiddle with these params no ? 

As I mentioned I'm actually interested in this working despite `{tuareg,caml}-mode`. The `ocaml` location format is good for printing indented text locations (which is nice for AST dumps or stack traces). The support in `compilation-mode` for the gnu format doesn't allow this.

> (Any reason for preferring caml-mode to tuareg?)

In the past, `caml-mode` was where the useful things got added first (e.g. `.annot` file support) but nowadays likely because I'm used to its syntax highlighting for the past 20 years.

(But if people get to eventually consolidate the two into an `ocaml-mode` I promise I won't scream)

**Comment by mattiase** [comment]
> Wasn't the idea to upstream the regexps ?

That wouldn't really solve any problem and even if it did, it would take years for changes to reach users.

>  In that case it would be better if they worked without having to fiddle with these params no ?

I agree that having buffer-local variables controlling how column numbers in compilation messages should be interpreted isn't ideal. For instance, if another tool emits a message referring to a location in OCaml source but using a different convention (gnu format, say), then that location will likely be off by one.

Fixing that bug would be easy enough (will do it in Tuareg later on if I can muster the strength), but if anyone cares about caml-mode then just setting the aforementioned variables is the easy solution for now.

> (But if people get to eventually consolidate the two into an ocaml-mode I promise I won't scream)

The ideal would indeed be a single mode that works, is actively maintained, with a predictable name ('ocaml-mode' would be fine).

**Comment by dbuenzli** [comment]
> That wouldn't really solve any problem and even if it did, it would take years for changes to reach users.

Unless you have a nice location format to suggest that deosn't need to start at the beginning of the line and that is already recognized by `compilation-mode` it absolutely does. 

At the moment if I want to use this format for non-ocaml tools I have to ask users to load `caml-mode` which, by side effect, installs the regexp. I'd rather have that built-in in `compilation-mode`.
